### ISSUE TITLE
Warn when a state value contains unsupported object type (add check on State.get())

### ISSUE THREAD

**Issue by parisholley** [avkonst/hookstate#96]
Easy to miss the "Limitations for state values" part of documentation, would be great if this threw a better exception.

```
const state = useState({data:new Map()});

state.data.get().has('foo');
```

```
Method Map.prototype.has called on incompatible receiver [object Object]
```

**Comment by avkonst** [comment]
Yes, but hard to achieve for 2 reasons:

1. performance impact as it has to do deep scan
2. there is no a generic way to identify plain objects (thanks to poor design of JavaScript). Maps and Sets - yes, but a user may have an instance of it's own class with pure properties....

However, you can come up with a plugin for such purpose, which would verify new values provided in set/merge and which could turn off automatically if not in production.

Here are the guidelines for writing a plugin: https://hookstate.js.org/docs/writing-plugin

If you end up with some plugin, feel free to share it with others by contributing into the repository.

Makes sense?

**Comment by avkonst** [comment]
Cross ref: https://github.com/avkonst/hookstate/issues/106

**Comment by avkonst** [comment]
This checking can be added under development mode on property get callback.

**Comment by toastal** [comment]
Shouldn't at least `ReadonlyMap` and  `ReadonlySet` work? They are valid types for many situations (such as `Map`s allowing objects/arrays/etc. as keys) and as `Readonly*` they won't be modified and could be skipped.

**Comment by avkonst** [comment]
Support for any type, including Date, Function, any YourType was implemented in Hookstate 4.0.0-rc7
